### PR TITLE
Added PDOStatement->fetchObject() stub

### PR DESCRIPTION
**Pull request by clxmstaab**
adding a PDO Stub as mentioned in the referenced issue.

closes https://github.com/vimeo/psalm/issues/4685

This turns unexpected error message of https://psalm.dev/r/6431cc5c64

from

> LessSpecificReturnStatement - 13:14 - The type 'false|object' is more general than the declared return type 'User' for User::load_by_id

into

> InvalidReturnStatement - test.php:16:16 - The inferred type 'User::class|false' does not match the declared return type 'User|false' for User::load_by_id (see https://psalm.dev/128)

so its obvious the new stubs are picked up, but it seems psalm is still not able to match the types as expected.
it doesn't realize that `User::class` is the same as `self` in this context?

my testfile:

```
<?php

class User {
    // Private properties
    private $id, $name;

    private function __construct () {}

    /**
     * @return false|self
     */
    public static function load_by_id ($id) {
        $pdo = new pdo("my-dsn");
        $stmt = $pdo->prepare('SELECT id, name FROM users WHERE id=?');
        $stmt->execute([$id]);
        return $stmt->fetchObject(__CLASS__);
    }

}

$user = User::load_by_id(1);
/** @psalm-trace $user */
```

### CODE REVIEW

**Comment by orklah** [comment]
I think this is an issue with the difference between `User` as the type that represent an instance of User Class and `User::class`, aka class-string<User> which is a string with the name of the class

Your stub should take that into account, you receive a class-string and return the actuel object

**Comment by staabm** [comment]
Good catch. I will try this tomorrow. 

Thx

**Comment by clxmstaab** [comment]
the suggestion from orklah works like a charm.

everything looks like expected now:

```
 cat test.php
<?php

class User {
    // Private properties
    private $id, $name;

    private function __construct () {}

    /**
     * @return false|self
     */
    public static function load_by_id ($id) {
        $pdo = new pdo("my-dsn");
        $stmt = $pdo->prepare('SELECT id, name FROM users WHERE id=?');
        $stmt->execute([$id]);
        return $stmt->fetchObject(__CLASS__);
    }

}

$user = User::load_by_id(1);
/** @psalm-trace $user */


php psalm test.php
Scanning files...
Analyzing files...

E

To whom it may concern: Psalm cannot detect unused classes, methods and properties
when analyzing individual files and folders. Run on the full project to enable
complete unused code detection.

ERROR: MissingPropertyType - test.php:5:13 - Property User::$id does not have a declared type (see https://psalm.dev/045)
    private $id, $name;


ERROR: MissingParamType - test.php:12:40 - Parameter $id has no provided type (see https://psalm.dev/154)
    public static function load_by_id ($id) {


ERROR: UnusedVariable - test.php:21:1 - $user is never referenced or the value is not used (see https://psalm.dev/077)
$user = User::load_by_id(1);


ERROR: Trace - test.php:22:0 - $user: User|false (see https://psalm.dev/224)
/** @psalm-trace $user */


------------------------------
4 errors found
------------------------------
Psalm can automatically fix 1 of these issues.
Run Psalm again with
--alter --issues=UnusedVariable --dry-run
to see what it can fix.
------------------------------

Checks took 1.74 seconds and used 160.969MB of memory
Psalm was unable to infer types in the codebase

```

the return type of `fetchObject` is correctly infered.

File: src/Psalm/Config.php [hunk 1]
```diff
@@ -1838,7 +1848,7 @@ public function visitStubFiles(Codebase $codebase, ?Progress $progress = null):
             $xdebug_stub_path = dirname(__DIR__, 2) . '/stubs/Xdebug.php';
 
             if (!file_exists($xdebug_stub_path)) {
-                throw new \UnexpectedValueException('Cannot locate XDebug stub');
+                throw new \UnexpectedValueException('Cannot locate Xdebug stub');
```
Comment (clxmstaab) [inline src/Psalm/Config.php:1851]: drive-by cleanup: the official name is Xdebug, see http://xdebug.org/

**Comment by muglug** [comment]
Thanks, this is great!

File: stubs/pdo.php [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+class PdoStatement {
+    /**
+     * @template T
+     * @param class-string<T> $class
+     * @return false|T
+     */
+    public function fetchObject($class = "stdclass") {}
```
Comment (staabm) [inline stubs/pdo.php:9]: @muglug should parameters like this one, which make the php engine create a object based on a userland class-name, be considered a taint-sink?

Comment (staabm) [reply stubs/pdo.php:9]: Just went ahead and proposed a change: https://github.com/vimeo/psalm/pull/4701
